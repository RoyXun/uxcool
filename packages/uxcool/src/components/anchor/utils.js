const REG_ANCHOR_ID = /#[^#]+$/;

export function getAnchorRelateElement(href = '') {
  const matches = href.match(REG_ANCHOR_ID);
  if (!matches) {
    return null;
  }

  return document.querySelector(matches[0]);
}

export function getElementOffsetTop(element, container) {
  let top = 0;
  if (!element || !container) {
    return top;
  }

  const box = element.getBoundingClientRect();
  if (container.window === window) {
    const doc = element.ownerDocument;
    const docElement = doc && doc.documentElement;
    const body = doc && doc.body;
    top = box.top - docElement.clientTop || body.clientTop || 0;
  } else {
    const containerRect = container.getBoundingClientRect();
    top = box.top - containerRect.top;
  }
  return top;
}
