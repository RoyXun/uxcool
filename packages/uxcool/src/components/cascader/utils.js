import { isFunction } from '@suning/v-utils';

export const SPLIT_STR = '\u0001';

export const DEFAULT_FIELD_NAMES = {
  label: 'label',
  value: 'value',
  children: 'children',
};

export function genCascaderKey(valuePath) {
  return valuePath.join(SPLIT_STR);
}

export function delay(fn, wait) {
  if (!isFunction(fn)) {
    return null;
  }
  return setTimeout(fn, wait);
}
