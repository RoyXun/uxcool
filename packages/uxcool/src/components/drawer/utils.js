import { isNaN, isNumber } from '@cloud-sn/v-utils';

export function genUid() {
  let dt = new Date().getTime();
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    // eslint-disable-next-line no-bitwise, no-mixed-operators
    const r = ((dt + Math.random() * 16) % 16) | 0;
    dt = Math.floor(dt / 16);
    // eslint-disable-next-line no-bitwise,
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });
}

export function noop() {}

export function isNumeric(value) {
  return !isNaN(value) && isNumber(value);
}

export function hasScrollBar() {
  return document.body.clientWidth < window.innerWidth;
}

export function getVisibleDrawerCnt(drawers = {}) {
  return Object.keys(drawers).filter((k) => drawers[k]).length;
}

export function hasVisibleDrawer(drawers = {}) {
  return getVisibleDrawerCnt(drawers) >= 1;
}

export function getDoc() {
  return window ? window.document.body : {};
}
