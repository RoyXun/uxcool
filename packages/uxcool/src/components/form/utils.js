import omit from 'object.omit';
import {
  warning,
  toArray,
  isFunction,
  isPlainObject,
  isString,
  setValueByPath,
  getValueByPath,
  isDef,
} from '@suning/v-utils';

import { Validator } from 'vee-validate';

const { hasOwnProperty } = Object.prototype;

export function isValidateField(field) {
  return field && isDef(field.rules);
}

export function getSlotOrValue(name, target) {
  const { $slots } = target;
  if ($slots[name]) {
    return $slots[name];
  }

  return target[name] || null;
}

export function hasPath(path, target) {
  let obj = target;
  return path.split('.').every((prop) => {
    if (!hasOwnProperty.call(obj, prop)) {
      return false;
    }
    obj = obj[prop];

    return true;
  });
}

// export const getPath = (path, target, def) => {
//   if (!path || !target) return def;

//   let value = target;
//   const paths = path.split('.');

//   for (let i = 0, l = paths.length; i < l; i += 1) {
//     const prop = paths[i];
//     if (!hasOwnProperty.call(value, prop) && value[prop] === undefined) {
//       value = def;
//       break;
//     }

//     value = value[prop];
//   }
//   return value;
// };

export function getPath(path, target, def) {
  return getValueByPath(target, path, def);
}
export function find(arrayLike, predicate) {
  const array = Array.isArray(arrayLike) ? arrayLike : toArray(arrayLike);
  for (let i = 0; i < array.length; i += 1) {
    if (predicate(array[i])) {
      return array[i];
    }
  }

  return undefined;
}

export function getModel(vnode) {
  return vnode.data
    ? vnode.data.model || find(vnode.data.directives, d => d.name === 'model')
    : null;
}

export function getVnodeOptions(vnode) {
  const component = vnode.componentInstance;
  const opts = {};
  return component
    ? {
      model: component.$options.model || {},
      listeners: component.$listeners || {},
    }
    : opts;
}

export function resolveModel(vnode) {
  const model = getModel(vnode);
  if (!model) {
    return null;
  }

  // https://github.com/vuejs/vue/blob/dev/src/core/util/lang.js#L26
  const watchable = !/[^\w.$]/.test(model.expression) && hasPath(model.expression, vnode.context);
  const lazy = !!(model.modifiers && model.modifiers.lazy);
  if (!watchable) {
    return {
      expression: null,
      lazy,
    };
  }

  return {
    expression: model.expression,
    lazy,
  };
}

export function resolveInitialValue(vnode) {
  const model = getModel(vnode);

  return model && model.value;
}

export function makeVM(vm, context) {
  return {
    get $el() {
      return vm.$el;
    },
    get $refs() {
      // 同理
      return context.$refs;
    },
    get context() {
      return context;
    },
    // $watch会观察model的expression,所以要用vnode.context
    $watch: context.$watch ? context.$watch.bind(context) : () => {},
    $validator: vm.$validator
      ? {
        errors: vm.$validator.errors,
        validate: vm.$validator.validate.bind(vm.$validator),
        update: vm.$validator.update.bind(vm.$validator),
      }
      : null,
  };
}

export function getCtorConfig(vnode) {
  if (!vnode.componentInstance) return null;

  const config = getPath('componentInstance.$options.$_veeValidate', vnode);

  return config;
}

export function resolveSetter(name, el, vnode) {
  const model = getModel(vnode);
  if (model && model.expression) {
    if (model.callback) {
      return model.callback;
    }

    if (model.expression && hasPath(vnode.context, model.expression)) {
      return val => setValueByPath(vnode.context, model.expression, val);
    }
  }

  const options = getVnodeOptions(vnode);
  const eventName = options.model.event || 'input';
  const listener = options.listeners[eventName];
  if (listener) {
    return val => vnode.componentInstance.$emit(eventName, val);
  }

  return () => {
    warning(
      false,
      `Please provide the \`v-model\` property or bind \`${eventName}\` events for the elements within the \`${name}\` FieldDecorator.`
    );
  };
}
export function resolveGetter(el, vnode, model, valuePath) {
  if (model && model.expression) {
    return () => getPath(model.expression, vnode.context);
  }

  if (valuePath) {
    return () => getPath(valuePath, vnode.context);
  }

  if (vnode.componentInstance) {
    const config = getCtorConfig(vnode);
    if (config && isFunction(config.value)) {
      const boundGetter = config.value.bind(vnode.componentInstance);

      return () => boundGetter();
    }

    return () => vnode.componentInstance.value;
  }

  switch (el.type) {
    case 'checkbox':
      return () => {
        let els = document.querySelectorAll(`input[name="${el.name}"]`);

        els = toArray(els).filter(v => v.checked);
        if (!els.length) return undefined;

        return els.map(checkbox => checkbox.value);
      };
    case 'radio':
      return () => {
        const els = document.querySelectorAll(`input[name="${el.name}"]`);
        const elm = find(els, v => v.checked);

        return elm && elm.value;
      };
    case 'file':
      return () => toArray(el.files);
    case 'select-multiple':
      return () =>
        toArray(el.options)
          .filter(opt => opt.selected)
          .map(opt => opt.value);
    default:
      return () => el && el.value;
  }
}

function parseRule(rule) {
  let params = [];
  const name = rule.split(':')[0];

  if (rule.indexOf(':') > -1) {
    params = rule
      .split(':')
      .slice(1)
      .join(':')
      .split(',');
    // 处理confirmed, after, before等规则
    if (Validator.isTargetRule(name)) {
      const selector = params[0];
      if (selector) {
        // @开头的selector 标示用户使用元素自定义属性
        // 由于元素存在嵌套的情况veeValidate的updateDependencies不能取到实际元素,导致无法获取实际值,所以让用户指定实际值
        if (selector.indexOf('@') !== 0) {
          params[0] = `$${selector.replace(/^\$/, '')}Ref`;
        } else {
          params[0] = selector.slice(1);
        }
      }
    }
  }
  return { name, params };
}
export function parseStringRules(rules) {
  return rules.split('|').reduce((p, rule) => {
    const np = p;
    const parsedRule = parseRule(rule);
    if (!parsedRule.name) {
      return np;
    }

    np[parsedRule.name] = parsedRule.params;
    return np;
  }, {});
}
export function parseFieldRules(rules) {
  let r = {
    rules: {},
    messages: {},
  };

  if (Array.isArray(rules)) {
    r = rules.reduce((p, v) => {
      const np = p;
      const rule = parseFieldRules(v);

      np.rules = { ...np.rules, ...rule.rules };
      np.messages = { ...np.messages, ...rule.messages };

      return np;
    }, r);
  } else if (isPlainObject(rules)) {
    if ('message' in rules) {
      const nRules = omit(rules, 'message');
      const keys = Object.keys(nRules);
      if (keys.length >= 1) {
        r = {
          rules: nRules,
          messages: {
            [keys[0]]: rules.message,
          },
        };
      }
    } else if (!('message' in rules)) {
      r.rules = rules;
    }
  } else if (isString(rules)) {
    r.rules = parseStringRules(rules);
  }

  return r;
}

export function validateField(field, validator) {
  if (field.initial) {
    validator.validate(`#${field.id}`, field.value);
  } else {
    // eslint-disable-next-line
    validator._validate(field, field.value, true).then(result => {
      field.setFlags({
        valid: result.valid,
        invalid: !result.valid,
      });
    });
  }
}
