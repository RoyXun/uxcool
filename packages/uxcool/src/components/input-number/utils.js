import isNaN from 'lodash/isNaN';

export { isNaN };

export function isNotCompleteNumber(v) {
  return (
    isNaN(Number(v)) ||
    v === '' ||
    v === null ||
    v === undefined ||
    (v && v.toString().slice('.') === v.length - 1)
  );
}

export function getPrecision(val, defaultPrecision = null) {
  if (defaultPrecision !== null) {
    return defaultPrecision;
  }

  let valStr = val.toString();
  let eDigit = 0;
  if (/e-/.test(valStr)) {
    [valStr, eDigit] = valStr.split('e-');
  }

  let decimal = 0;
  if (/\./.test(valStr)) {
    decimal = parseInt(valStr.split('.')[1].length, 10);
  }

  return decimal + eDigit;
}

export function getMaxPrecision(val, step, ratio = 1, defaultPrecision = null) {
  if (defaultPrecision !== null) {
    return defaultPrecision;
  }

  const stepPrecision = getPrecision(step);
  const ratioPrecision = getPrecision(ratio);
  const valPrecision = getPrecision(val);

  return Math.max(valPrecision, stepPrecision + ratioPrecision);
}

export function getRatio(e) {
  let ratio = 1;
  if (e.metaKey || e.ctrlKey) {
    ratio = 0.1;
  } else if (e.shiftKey) {
    ratio = 10;
  }
  return ratio;
}
