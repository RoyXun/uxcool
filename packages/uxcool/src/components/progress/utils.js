export function defaultInfoFormat(percent) {
  return `${percent}%`;
}

export function getCirclePath(pos, radius) {
  const { start, end } = pos;
  return `M 50 50 m ${start.x},${start.y}
          a ${radius},${radius} 0 1 1 ${end.x},${end.y}
          a ${radius},${radius} 0 1 1 ${-1 * end.x},${-1 * end.y}`;
}

export function getPosition(gapPos, radius) {
  let start = {
    x: 0,
    y: -radius,
  };
  let end = {
    x: 0,
    y: 2 * radius,
  };

  switch (gapPos) {
    case 'left':
      start = {
        x: -radius,
        y: 0,
      };
      end = {
        x: 2 * radius,
        y: 0,
      };
      break;
    case 'right':
      start = { x: radius, y: 0 };
      end = {
        x: -2 * radius,
        y: 0,
      };
      break;
    case 'bottom':
      start = {
        x: 0,
        y: radius,
      };
      end = {
        x: 0,
        y: -2 * radius,
      };
      break;
    default:
      break;
  }

  return {
    start,
    end,
  };
}
