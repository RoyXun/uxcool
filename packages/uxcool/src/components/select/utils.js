import {
  isSameTypeVNode,
  isArray,
  cloneVNode,
  isDef,
  isFunction,
  warning,
  extractVNodeData,
  isVueComponent,
} from '@cloud-sn/v-utils';

export function isValidValue(value) {
  return isDef(value) && value !== '';
}
export function buildOptionsFromSlot(nodes = []) {
  const ret = [];

  for (let i = 0, l = nodes.length; i < l; i += 1) {
    const node = nodes[i];
    if (isSameTypeVNode(node, 'isOptionGroupType')) {
      const { componentOptions: cp, data: { attrs } } = node;
      const props = cp.propsData;
      if (process.env.NODE_ENV !== 'production' && !isDef(props.label)) {
        warning(
          false,
          'OptionGroup invalid prop: type check failed for prop "label". Expected String,got undefined .'
        );
      }
      ret.push({
        ...attrs,
        ...props,
        label: props.label,
        children: buildOptionsFromSlot(cp.children),
      });
    } else if (isSameTypeVNode(node, 'isOptionType')) {
      const { componentOptions: cp, data: { attrs } } = node;
      const props = cp.propsData;
      if (process.env.NODE_ENV !== 'production' && !isDef(props.value)) {
        warning(
          false,
          'Option invalid prop: type check failed for prop "value". Expected String, Number ,got undefined .'
        );
      }
      ret.push({
        ...attrs,
        ...props,
        value: props.value,
        label: props.label || '',
        labelNode: cp.children,
        disabled: 'disabled' in props && props.disabled !== false,
      });
    }
  }
  return ret;
}

export function isUnRenderOption(option) {
  return 'UNRENDER_OTPION' in option && option.UNRENDER_OTPION === true;
}
export function isOptionGroup(option) {
  return 'children' in option;
}

export function cloneLabelNodes(vnodes) {
  return isArray(vnodes) ? vnodes.map((node) => cloneVNode(node)) : cloneVNode(vnodes);
}

export function normalizeContent(maxLen, content) {
  const l = Number(maxLen);
  if (l <= 0 || typeof content !== 'string' || content.length < l) {
    return content;
  }

  return `${content.slice(0, l)}...`;
}

export function DefaultFilterOption(filterProp) {
  return (inputValue, option) => {
    if (option.disabled) {
      return false;
    }
    const optionValue = String(option[filterProp]) || '';
    return optionValue.toLowerCase().indexOf(inputValue.toLowerCase()) > -1;
  };
}

export function recursiveSearch(options = [], inputValue, filterFn) {
  const ret = [];
  for (let i = 0, l = options.length; i < l; i += 1) {
    const option = options[i];
    if (isOptionGroup(option)) {
      const { children } = option;
      const nChildren = recursiveSearch(children || [], inputValue, filterFn);
      if (nChildren.length > 0) {
        ret.push({ ...option, children: nChildren });
      }
    } else if (filterFn(inputValue, option.originNode)) {
      ret.push(option);
    }
  }
  return ret;
}

export function buildOptionOriginNode(name, optionMap) {
  if (!isDef(name)) {
    return {};
  }
  if (name in optionMap) {
    const { originNode } = optionMap[name];
    const {
      label, value, title, disabled
    } = originNode;
    const nLabel = label || value;
    return {
      ...originNode,
      disabled: !!disabled,
      title: title || nLabel,
      label: nLabel,
      value,
    };
  }
  return {
    disabled: false,
    title: name,
    label: name,
    value: name,
  };
}

export function calcMultipleSearchInputWidth(inputValue, inputMirror) {
  if (inputValue && inputMirror) {
    return {
      width: `${inputMirror.clientWidth}px`,
    };
  }
  return {};
}

function getRegByTokens(tokens = []) {
  return new RegExp(`^[^${tokens.join('')}]+[${tokens.join('')}]`);
}
export function includeSeparators(value = '', tokens = []) {
  const reg = getRegByTokens(tokens);
  return reg.test(value);
}

export function getOptionLabelMap(optionMap = {}) {
  return Object.keys(optionMap).reduce((r, k) => {
    const nr = r;
    const option = optionMap[k];
    nr[option.label || k] = option;
    return nr;
  }, {});
}

export function splitValueBySeparator(value = '', tokens = []) {
  const splitReg = new RegExp(`[${tokens.join('')}]`);
  return value.split(splitReg).filter((v) => !!v);
}

export function getComboboxValue(value = '', optionMap = {}, optionLabelProp = 'value') {
  const comboboxValue = isArray(value) ? value[0] : value;
  let label = '';
  if (comboboxValue) {
    if (optionLabelProp === 'value') {
      label = comboboxValue;
    } else {
      // optionLabelProp = label;
      const option = optionMap[comboboxValue];
      if (option) {
        label = option.label; //eslint-disable-line
      } else {
        label = comboboxValue;
      }
    }
  }
  return label;
}

export function getOptionOriginNode(optionMap = {}) {
  return Object.keys(optionMap).map((k) => {
    const option = optionMap[k];
    return option.originNode;
  });
}

export function genInputElement(inputNode, data = {}) {
  const overlayData = extractVNodeData(inputNode, isVueComponent(inputNode));

  const { input: originInput } = overlayData.on || {};
  const { input: onInput } = data.on || {};
  const inputFn = isFunction(onInput) ? (_, e) => onInput(e) : () => {};
  const inputInput = originInput ? [originInput, inputFn] : inputFn;
  const ref = data.ref || inputNode.data.ref;
  return cloneVNode(
    inputNode,
    {
      class: [overlayData.class, data.class],
      props: {
        ...overlayData.props,
        ...data.props,
      },
      on: {
        ...overlayData.on,
        input: inputInput,
      },
      ref,
    },
    false
  );
}
