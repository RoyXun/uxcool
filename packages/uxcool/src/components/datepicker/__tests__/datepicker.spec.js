import { mount, waitTime } from '@suning/v-test-utils';
import dayjs from 'dayjs';
import { selectDate, getPortal, triggerEvent } from './utils';
import DatePicker from '../datepicker.vue';

describe('DatePicker', () => {
  it('create', () => {
    const format = 'YYYY-MM-DD';
    const placeholder = 'create-test';
    const id = 'test-a';
    const wrapper = mount(DatePicker, {
      propsData: {
        format,
        placeholder,
      },
      attrs: {
        id,
      },
    });

    const input = wrapper.find('input');

    expect(input.attributes('placeholder')).toBe(placeholder);
    expect(wrapper.element.id).toBe(id);
    // const selectedDate = formatDate(subDays(new Date(), 1), format);
    // const input = wrapper.find('input');
    // input.trigger('click');

    // await waitTime(20);
    // const triggerWrapper = wrapper.find(Trigger);
    // const portalWrapper = createWrapper(triggerWrapper.vm.portal.$el);
    // console.log('portalWrapper, %o', portalWrapper);
    // const a = portalWrapper.find(`[role="gridcell"][title="${selectedDate}"]`);
    // console.log('a%o', a);
    // portalWrapper.find(`[role="gridcell"][title="${selectedDate}"]`).trigger('click');

    // expect(wrapper.find('input').text()).toBe(selectedDate);
  });

  it('select date', async () => {
    const format = 'YYYY-MM-DD';
    const wrapper = mount(DatePicker, {
      propsData: {
        format,
        // isOpen: true,
      },
    });
    // wait render
    await waitTime(20);

    const inputWrapper = wrapper.find('.ux-input');
    await triggerEvent(wrapper, 'click');
    const portalWrapper = await getPortal(wrapper);

    const prevDate = dayjs()
      .subtract(1, 'date')
      .format(format);
    // const prevDate = formatDate(subDays(new Date(), 1), format);
    await selectDate(portalWrapper, prevDate);
    expect(inputWrapper.element.value).toBe(prevDate);

    const prevMonth = dayjs()
      .subtract(1, 'month')
      .format(format);
    //  formatDate(subMonths(new Date(), 1), format);
    await selectDate(portalWrapper, prevMonth);
    expect(inputWrapper.element.value).toBe(prevMonth);
  });

  it('is-open', async () => {
    const wrapper = mount(DatePicker, {
      propsData: {
        isOpen: true,
      },
    });
    await waitTime(20);

    const portal = await getPortal(wrapper);
    expect(portal.find('.ux-calendar-picker-container').exists()).toBeTruthy();

    wrapper.setProps({
      isOpen: false,
    });
    // animate delay
    await waitTime(300);
    expect(portal.find('.ux-calendar-picker-container').exists()).toBeFalsy();

    wrapper.setProps({
      isOpen: true,
    });
    // animate delay
    await waitTime(300);
    expect(portal.find('.ux-calendar-picker-container').exists()).toBeTruthy();
  });

  it('value(v-model) no transitionName', async () => {
    const format = 'YYYY-MM-DD';
    const Test = {
      data() {
        return {
          value: null,
        };
      },
      methods: {
        onInput(value) {
          this.value = value;
        },
      },
      render() {
        const { value, onInput } = this;
        return <DatePicker value={value} transitionName="" on-input={onInput} />;
      },
    };

    const wrapper = mount(Test);
    const input = wrapper.find('.ux-input');
    await waitTime(20);
    const dp = wrapper.find(DatePicker);
    await triggerEvent(dp, 'click');
    const portal = await getPortal(wrapper);

    await triggerEvent(portal.find('.ux-calendar-next-month-btn'), 'click');
    const nextMonth = dayjs()
      .add(1, 'month')
      .format('format');
    // const nextMonth = formatDate(addMonths(new Date(), 1), format);
    await selectDate(portal, nextMonth);
    expect(dayjs(wrapper.vm.value).format(format)).toBe(nextMonth);
    expect(input.element.value).toBe(nextMonth);

    await triggerEvent(dp, 'click');
    const prevMonthBtn = portal.find('.ux-calendar-prev-month-btn');
    await triggerEvent(prevMonthBtn, 'click');
    await triggerEvent(prevMonthBtn, 'click');
    const prevMonth = dayjs()
      .subtract(1, 'month')
      .format(format);
    // formatDate(subMonths(new Date(), 1), format);
    // const prevMonth = formatDate(subMonths(new Date(), 1), format);
    await selectDate(portal, prevMonth);
    expect(dayjs(wrapper.vm.value).format(format)).toBe(prevMonth);
    expect(input.element.value).toBe(prevMonth);

    await triggerEvent(dp, 'click');

    const nDate = dayjs().subtract(1, 'year');
    wrapper.setData({
      value: nDate,
    });
    await waitTime(20);

    const selectedElement = portal.find('.ux-calendar-selected-date');
    expect(selectedElement.exists()).toBeTruthy();

    expect(selectedElement.attributes('title')).toBe(dayjs(nDate).format(format));
  });

  it('disabled', async () => {
    const wrapper = mount(DatePicker, {
      propsData: {
        disabled: true,
      },
    });
    await waitTime(20);
    expect(wrapper.find('.ux-input-disabled').exists()).toBeTruthy();

    wrapper.setProps({
      disabled: false,
    });
    await waitTime(20);
    expect(wrapper.find('.ux-input-disabled').exists()).toBeFalsy();
  });

  it('format DD/MM/YYYY', async () => {
    const format = 'DD/MM/YYYY';
    const today = new Date();
    const wrapper = mount(DatePicker, {
      propsData: {
        value: today,
        format,
        transitionName: '',
      },
    });

    await waitTime(20);

    const input = wrapper.find('.ux-input');
    expect(input.element.value).toBe(dayjs(today).format(format));

    await triggerEvent(wrapper, 'click');

    const portal = await getPortal(wrapper);
    const nDate = dayjs()
      .add(2, 'date')
      .format(format);
    // formatDate(addDays(new Date(), 2), format);
    await selectDate(portal, nDate);
    expect(input.element.value).toBe(nDate);
  });

  it.only('show-time', async () => {
    jest.setTimeout(50000);
    const wrapper = mount(DatePicker, {
      propsData: {
        showTime: true,
        transitionName: '',
      },
    });

    await waitTime(20);

    await triggerEvent(wrapper, 'click');

    const portal = await getPortal(wrapper);

    const timePickerBtn = portal.find('.ux-calendar-time-picker-btn');
    const okBtn = portal.find('.ux-calendar-ok-btn');
    expect(timePickerBtn.exists()).toBeTruthy();
    expect(okBtn.exists()).toBeTruthy();

    await triggerEvent(timePickerBtn, 'click');
    const timerPicker = portal.find('.ux-calendar-time-picker');
    expect(timerPicker.exists()).toBeTruthy();

    wrapper.setProps({
      showTime: false,
    });
    await waitTime(20);

    expect(timePickerBtn.exists()).toBeFalsy();
    expect(okBtn.exists()).toBeFalsy();
    expect(timerPicker.exists()).toBeFalsy();
  });
});
