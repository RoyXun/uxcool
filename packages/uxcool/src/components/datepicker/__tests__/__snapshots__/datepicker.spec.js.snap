// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker locale 1`] = `
<div>
  <div class="ux-calendar-picker ux-calendar-picker-light ux-calendar-picker-default" size="default"><input placeholder="Select date" type="text" readonly="readonly" class="ux-input ux-calendar-picker-input"> <i class="fu fu-close_circle ux-calendar-picker-clear"></i> <span class="ux-calendar-picker-icon"></span></div>
  <div role="popup">
    <div role="align-popup" class="ux-calendar-picker-container ux-calendar-picker-container-placement-bottomLeft " style="display: none;">
      <div class="ux-calendar-picker-container-content">
        <div class="ux-calendar ux-calendar-light">
          <div role="panel" class="ux-calendar-panel">
            <div class="ux-calendar-input-wrap">
              <div class="ux-calendar-date-input-wrap"><input title="2019-10-22" placeholder="Select date" readonly="readonly" class="ux-calendar-input"></div>
              <!---->
            </div>
            <div class="ux-calendar-date-panel">
              <div class="ux-calendar-header">
                <div style="position: relative;"><a title="Last year (Control + left)" role="button" class="ux-calendar-prev-year-btn"></a> <a title="Previous month (PageUp)" role="button" class="ux-calendar-prev-month-btn"></a> <span class="ux-calendar-my-select"><a role="button" title="Choose a month" class="ux-calendar-month-select">Oct</a><a role="button" title="Choose a year" class="ux-calendar-year-select">2019</a></span> <a title="Next month (PageDown)" role="button" class="ux-calendar-next-month-btn"></a> <a title="Next year (Control + right)" role="button" class="ux-calendar-next-year-btn"></a></div>
                <div>
                  <!---->
                  <!---->
                  <!---->
                </div>
              </div>
              <!---->
              <div class="ux-calendar-body">
                <table cellspacing="0" role="grid" class="ux-calendar-table">
                  <thead>
                    <tr role="row">
                      <th title="Sunday" role="columnheader" class="ux-calendar-column-header"><span class="ux-calendar-column-header-inner">
        Su
      </span></th>
                      <th title="Monday" role="columnheader" class="ux-calendar-column-header"><span class="ux-calendar-column-header-inner">
        Mo
      </span></th>
                      <th title="Tuesday" role="columnheader" class="ux-calendar-column-header"><span class="ux-calendar-column-header-inner">
        Tu
      </span></th>
                      <th title="Wednesday" role="columnheader" class="ux-calendar-column-header"><span class="ux-calendar-column-header-inner">
        We
      </span></th>
                      <th title="Thursday" role="columnheader" class="ux-calendar-column-header"><span class="ux-calendar-column-header-inner">
        Th
      </span></th>
                      <th title="Friday" role="columnheader" class="ux-calendar-column-header"><span class="ux-calendar-column-header-inner">
        Fr
      </span></th>
                      <th title="Saturday" role="columnheader" class="ux-calendar-column-header"><span class="ux-calendar-column-header-inner">
        Sa
      </span></th>
                    </tr>
                  </thead>
                  <tbody class="ux-calendar-tbody" range-values="">
                    <tr>
                      <td role="gridcell" title="2019-09-29" class="ux-calendar-cell ux-calendar-last-month-cell">
                        <div class="ux-calendar-date">29</div>
                      </td>
                      <td role="gridcell" title="2019-09-30" class="ux-calendar-cell ux-calendar-last-month-cell">
                        <div class="ux-calendar-date">30</div>
                      </td>
                      <td role="gridcell" title="2019-10-01" class="ux-calendar-cell">
                        <div class="ux-calendar-date">1</div>
                      </td>
                      <td role="gridcell" title="2019-10-02" class="ux-calendar-cell">
                        <div class="ux-calendar-date">2</div>
                      </td>
                      <td role="gridcell" title="2019-10-03" class="ux-calendar-cell">
                        <div class="ux-calendar-date">3</div>
                      </td>
                      <td role="gridcell" title="2019-10-04" class="ux-calendar-cell">
                        <div class="ux-calendar-date">4</div>
                      </td>
                      <td role="gridcell" title="2019-10-05" class="ux-calendar-cell">
                        <div class="ux-calendar-date">5</div>
                      </td>
                    </tr>
                    <tr>
                      <td role="gridcell" title="2019-10-06" class="ux-calendar-cell">
                        <div class="ux-calendar-date">6</div>
                      </td>
                      <td role="gridcell" title="2019-10-07" class="ux-calendar-cell">
                        <div class="ux-calendar-date">7</div>
                      </td>
                      <td role="gridcell" title="2019-10-08" class="ux-calendar-cell">
                        <div class="ux-calendar-date">8</div>
                      </td>
                      <td role="gridcell" title="2019-10-09" class="ux-calendar-cell">
                        <div class="ux-calendar-date">9</div>
                      </td>
                      <td role="gridcell" title="2019-10-10" class="ux-calendar-cell">
                        <div class="ux-calendar-date">10</div>
                      </td>
                      <td role="gridcell" title="2019-10-11" class="ux-calendar-cell">
                        <div class="ux-calendar-date">11</div>
                      </td>
                      <td role="gridcell" title="2019-10-12" class="ux-calendar-cell">
                        <div class="ux-calendar-date">12</div>
                      </td>
                    </tr>
                    <tr>
                      <td role="gridcell" title="2019-10-13" class="ux-calendar-cell">
                        <div class="ux-calendar-date">13</div>
                      </td>
                      <td role="gridcell" title="2019-10-14" class="ux-calendar-cell">
                        <div class="ux-calendar-date">14</div>
                      </td>
                      <td role="gridcell" title="2019-10-15" class="ux-calendar-cell">
                        <div class="ux-calendar-date">15</div>
                      </td>
                      <td role="gridcell" title="2019-10-16" class="ux-calendar-cell">
                        <div class="ux-calendar-date">16</div>
                      </td>
                      <td role="gridcell" title="2019-10-17" class="ux-calendar-cell">
                        <div class="ux-calendar-date">17</div>
                      </td>
                      <td role="gridcell" title="2019-10-18" class="ux-calendar-cell">
                        <div class="ux-calendar-date">18</div>
                      </td>
                      <td role="gridcell" title="2019-10-19" class="ux-calendar-cell">
                        <div class="ux-calendar-date">19</div>
                      </td>
                    </tr>
                    <tr>
                      <td role="gridcell" title="2019-10-20" class="ux-calendar-cell">
                        <div class="ux-calendar-date">20</div>
                      </td>
                      <td role="gridcell" title="2019-10-21" class="ux-calendar-cell">
                        <div class="ux-calendar-date">21</div>
                      </td>
                      <td role="gridcell" title="2019-10-22" class="ux-calendar-cell ux-calendar-selected-date ux-calendar-selected-day">
                        <div class="ux-calendar-date">22</div>
                      </td>
                      <td role="gridcell" title="2019-10-23" class="ux-calendar-cell">
                        <div class="ux-calendar-date">23</div>
                      </td>
                      <td role="gridcell" title="2019-10-24" class="ux-calendar-cell">
                        <div class="ux-calendar-date">24</div>
                      </td>
                      <td role="gridcell" title="2019-10-25" class="ux-calendar-cell">
                        <div class="ux-calendar-date">25</div>
                      </td>
                      <td role="gridcell" title="2019-10-26" class="ux-calendar-cell">
                        <div class="ux-calendar-date">26</div>
                      </td>
                    </tr>
                    <tr>
                      <td role="gridcell" title="2019-10-27" class="ux-calendar-cell">
                        <div class="ux-calendar-date">27</div>
                      </td>
                      <td role="gridcell" title="2019-10-28" class="ux-calendar-cell">
                        <div class="ux-calendar-date">28</div>
                      </td>
                      <td role="gridcell" title="2019-10-29" class="ux-calendar-cell">
                        <div class="ux-calendar-date">29</div>
                      </td>
                      <td role="gridcell" title="2019-10-30" class="ux-calendar-cell">
                        <div class="ux-calendar-date">30</div>
                      </td>
                      <td role="gridcell" title="2019-10-31" class="ux-calendar-cell">
                        <div class="ux-calendar-date">31</div>
                      </td>
                      <td role="gridcell" title="2019-11-01" class="ux-calendar-cell ux-calendar-next-month-btn-day">
                        <div class="ux-calendar-date">1</div>
                      </td>
                      <td role="gridcell" title="2019-11-02" class="ux-calendar-cell ux-calendar-next-month-btn-day">
                        <div class="ux-calendar-date">2</div>
                      </td>
                    </tr>
                    <tr>
                      <td role="gridcell" title="2019-11-03" class="ux-calendar-cell ux-calendar-next-month-btn-day">
                        <div class="ux-calendar-date">3</div>
                      </td>
                      <td role="gridcell" title="2019-11-04" class="ux-calendar-cell ux-calendar-next-month-btn-day">
                        <div class="ux-calendar-date">4</div>
                      </td>
                      <td role="gridcell" title="2019-11-05" class="ux-calendar-cell ux-calendar-next-month-btn-day">
                        <div class="ux-calendar-date">5</div>
                      </td>
                      <td role="gridcell" title="2019-11-06" class="ux-calendar-cell ux-calendar-next-month-btn-day">
                        <div class="ux-calendar-date">6</div>
                      </td>
                      <td role="gridcell" title="2019-11-07" class="ux-calendar-cell ux-calendar-next-month-btn-day">
                        <div class="ux-calendar-date">7</div>
                      </td>
                      <td role="gridcell" title="2019-11-08" class="ux-calendar-cell ux-calendar-next-month-btn-day">
                        <div class="ux-calendar-date">8</div>
                      </td>
                      <td role="gridcell" title="2019-11-09" class="ux-calendar-cell ux-calendar-next-month-btn-day">
                        <div class="ux-calendar-date">9</div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
              <div class="ux-calendar-footer" style="display: none;"><span class="ux-calendar-footer-btn"><!----> <!----> <!----></span></div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
